packages/access-client/src/agent-data.js

```diff
@@ -0,0 +1,160 @@
+import { Signer } from '@ucanto/principal'
+import { Signer as EdSigner } from '@ucanto/principal/ed25519'
+import { importDAG } from '@ucanto/core/delegation'
+import { CID } from 'multiformats'
+
+/** @typedef {import('./types').AgentDataModel} AgentDataModel */
+
+/**
+ * @implements {AgentDataModel}
+ */
+export class AgentData {
+  /** @type {(data: import('./types').AgentDataExport) => Promise<void> | void} */
+  #save
+
+  /**
+   * @param {import('./types').AgentDataModel} data
+   * @param {import('./types').AgentDataOptions} [options]
+   */
+  constructor(data, options = {}) {
+    this.meta = data.meta
+    this.principal = data.principal
+    this.spaces = data.spaces
+    this.delegations = data.delegations
+    this.currentSpace = data.currentSpace
+    this.#save = options.store ? options.store.save : () => {}
+  }
+
+  /**
+   * @param {Partial<import('./types').AgentDataModel>} [init]
+   * @param {import('./types').AgentDataOptions} [options]
+   */
+  static async create(init = {}, options = {}) {
+    const agentData = new AgentData(
+      {
+        meta: { name: 'agent', type: 'device', ...init.meta },
+        principal: init.principal ?? (await EdSigner.generate()),
+        spaces: init.spaces ?? new Map(),
+        delegations: init.delegations ?? new Map(),
+        currentSpace: init.currentSpace,
+      },
+      options
+    )
+    if (options.store) {
+      await options.store.save(agentData.export())
+    }
+    return agentData
+  }
+
+  /**
+   * Instantiate AgentData, backed by data persisted in the passed store.
+   *
+   * @param {import('./types').IStore<import('./types').AgentDataExport>} store
+   * @param {import('./types').AgentDataOptions & { initialData?: Partial<import('./types').AgentDataModel> }} options
+   */
+  static async fromStore(store, options = {}) {
+    await store.open()
+    const storedData = await store.load() // { ... } or null/undefined
+    return storedData
+      ? AgentData.fromExport(storedData)
+      : await AgentData.create(options.initialData, {
+          store: {
+            save: async (d) => {
+              await store.save(d)
+            },
+          },
+        })
+  }
```

Comment:
if we dont have a method to test if the store was initialized do we always call this method ? if yes we gonna need to always pass on a Signer just in case its empty ?

Reply:
> if we dont have a method to test if the store was initialized do we always call this method ?

Yes, this is probably what you use 99% of the time. Essentially `store.load()` is the init check (which is what is happening above).

> if yes we gonna need to always pass on a Signer just in case its empty ?

No because `AgentData.create` generates one for you. I understand that's not great if you want to use a different key type from what `AgentData` gives you by default...and in that situation you'd either always pass it (as you say), or do what this method is doing in your application code. I don't know how to make that a better experience other than picking a better default in the browser.

Reply:
can we make the store.load() do store.open() ?

Reply:
humm so user needs to know about:
- Agent
- AgentData
- StoreDriver
- Signer

We can leave Signer in the options for flexibility, but right now BrowserDriver kinda requires RSASigner.

can we reduce this to:
- Agent
- Driver
- Signer (this would be optional, either from the driver or from the user overriding the driver)

Most uses would just need Agent and Driver, i think we can do this by just hiding AgentData inside Agent.

Reply:
```js

const data = driver.load()

if(data) {
  agent.from(data, driver)
}else{
  agent.create(driver)
}
```

something like this ?

Reply:
> ```js
> const data = driver.load()
> 
> if(data) {
>   agent.from(data, driver)
> }else{
>   agent.create(driver, { 
>     principal: await EdSigner.generate() // if we want to overriding Driver.generatePrincipal() 
>   })
> }
> ```
> 
> something like this ?

yes we can do that